research/picai/preprocessing.py

```diff
@@ -0,0 +1,358 @@
+        """
+        self.scan_paths = scan_paths
+        self.annotations_path = annotations_path
+        self.settings = settings
+
+        self.scans: List[sitk.Image]
+        self.annotation: sitk.Image
+
+    @abstractmethod
+    def read(self) -> None:
+        """
+        Abstract method to be implemented by children that reads the preprocessed scans and annotation
+        into memory.
+
+        Raises:
+            NotImplementedError
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def write(self) -> Tuple[Sequence[Path], Path]:
+        """
+        Abstract method to be implemented by children that writes the preprocessed scans and annotation
+        to their destination and returns the file paths.
+
+        Returns:
+            Tuple[Sequence[Path], Path]: A tuple in which the first entry is a sequence of file paths
+                for the scans and the second entry is the file path to the correpsonding annotation.
+
+        Raises:
+            NotImplementedError
+        """
+        raise NotImplementedError
+
+
+class PicaiCase(Case):
+    def __init__(
+        self,
+        scan_paths: Sequence[Path],
+        annotations_path: Path,
+        settings: PreprocessingSettings,
+    ) -> None:
+        """
+        Class representing a case from the PICAI dataset.
+
+        Args:
+            scan_paths (Sequence[Path]): The set of paths where the scans associated with the Case are located.
+            annotation_write_dir (Path): The path where the annotation associated with the Case is located.
+            settings (PreprocessingSettings): The settings determining how the case is preprocessed.
+        """
+        super().__init__(scan_paths, annotations_path, settings)
+
+    def read(self) -> None:
+        """
+        Reads in scans and annotation into the corresponding scans and annotation attribute.
+        """
+        assert len(self.scan_paths) != 0
+        self.scans = [sitk.ReadImage(path) for path in self.scan_paths]
+        self.annotation = sitk.ReadImage(self.annotations_path)
+
+    def write(self) -> Tuple[Sequence[Path], Path]:
+        """
+        Writes preprocessed scans and annotations from PICAI dataset to disk
+        and returns the scan file paths and annotation file path in a tuple.
+
+        Returns:
+            Tuple[Sequence[Path], Path]: A tuple in which the first entry is a sequence of file paths
+                for the scans and the second entry is the file path to the correpsonding annotation.
+        """
+        modality_suffix_map = {"t2w": "0000", "adc": "0001", "hbv": "0002"}
+        scan_paths = [path for path in sorted(self.scan_paths)]
+        preprocessed_scan_paths = []
+        for path, scan in zip(scan_paths, self.scans):
+            scan_filename = os.path.basename(path)
+            scan_filename_without_extension = scan_filename.split(".")[0]
```

Comment:
I think Path objects have a function that does this for you, if I recall correctly.